bin/install-core

```diff
@@ -120,13 +124,38 @@ mkdir -p "$DOWNLOAD_DIR"
 
 # TODO: uncomment the downloads when done testing other things
 # Download Bitcoin core checksums and signatures in background
-# TODO: these are messy make them legible.
-( cd "$DOWNLOAD_DIR"; #wget --tries=1 --no-host-directories --cut-dirs=1 --recursive --level=1 --accept=SHA256SUMS* --include-directories=bin http://6hasakffvppilxgehrswmffqurlcjjjhd76jgvaqmsg6ul25s7t3rzyd.onion/en/download/ ||
-wget --no-host-directories --cut-dirs=1 --recursive --level=1 --accept=SHA256SUMS* --include-directories=bin https://bitcoincore.org/en/download/; ) & sig_dl=$!
+(
+    cd "$DOWNLOAD_DIR"
+    for domain in ${BITCOIN_CORE_SOURCE_DOMAINS[*]}; do
+        wget --tries=1 --no-host-directories --cut-dirs=1 --recursive \
+            --level=1 --accept=SHA256SUMS* --include-directories=bin \
+            "https://${domain}/en/download/" && break
+    done
+
+    if [ $? -ne 0 ]; then
+        echo "Failed to download Bitcoin Core verification files."
+        exit 1
+    fi
+) & sig_dl=$!
 # TODO FIXME uncomment the onion service when bitcoincore.org fixes their service.
 # Download Bitcoin Core in background to the destination directory
-( sleep 1; while pgrep wget &>/dev/null; do sleep 1; done; cd "$DOWNLOAD_DIR"; #wget --continue --tries=1 --no-host-directories --cut-dirs=1 --recursive --level=1 --accept=x86_64-linux-gnu.tar.gz --include-directories=bin http://6hasakffvppilxgehrswmffqurlcjjjhd76jgvaqmsg6ul25s7t3rzyd.onion/en/download/ ||
-wget --continue --no-host-directories --cut-dirs=1 --recursive --level=1 --accept=x86_64-linux-gnu.tar.gz --include-directories=bin https://bitcoincore.org/en/download/; ) & core_download=$!
+(
+    sleep 1
+    while pgrep wget &>/dev/null; do
+        sleep 1
+    done
+    cd "$DOWNLOAD_DIR"
+    for domain in ${BITCOIN_CORE_SOURCE_DOMAINS[*]}; do
+        wget --continue --tries=1 --no-host-directories --cut-dirs=1 --recursive \
+            --level=1 --accept=x86_64-linux-gnu.tar.gz --include-directories=bin \
+            "http://${domain}/en/download/" && break
+    done
+
+    if [ $? -ne 0 ]; then
+        echo "Failed to download the Bitcoin Core package."
+        exit 1
```

Comment:
Exit in a sub-shell will not exit the entire script. `() &` makes the entire block within execute in a subshell.

Reply:
To exit the script we need to check the exit status of the PID stored in `$core_download` when we get to the point of waiting on it, and if it was 1, exit 1 there (as well) in the main thread.

Reply:
TODO

Reply:
Does this mean that the check here can be removed? Or should I set the $core_download exit status to 1?

Reply:
No. It means these exit conditions are in the wrong spot.

![image](https://github.com/BenWestgate/Bails/assets/73506583/4dea605a-79c1-4eae-8dd7-651c3a9d3442)

If the exit status of the `$core_download` is 1, that's when the message should be displayed and the script exited.
Likewise for the signatures. Because the download happens in a background thread, putting the exit inside the thread only exits that background thread not the entire script.

Would you like me to fix it for you? And then will it be ready to merge if that's the only blocking issue?

Reply:
alternatively the child process could kill its parent process when download failure occurs. That's a bit less graceful as the parent could be in the middle of an operation that creates corrupted files if killed during.

Reply:
Yes, to do what you want this check must be in the main script or it must kill its parent. I would move it to where we wait for the process to finish.

If the last operation of the sub-shell failed, which is currently wget the exit status of the `$core_download` process will also be 1 and we can check `$?` after we wait for it and exit the script with this message there.

The exit status of `wait` will be the exit status of the process it was waiting for.

Also the variable names should be changed to be consistent. Either both spelling out `download` or both abbreviated `dl` and spelling it out is best.

Reply:
> @epiccurious: the script already had checks about whether the download worked or not.

source: https://github.com/BenWestgate/Bails/pull/94#discussion_r1592433559

@BenWestgate is this resolved since we're already doing file checks?

Reply:
Yes.